teps/0119-add-taskrun-template-in-pipelinerun.md

```diff
@@ -0,0 +1,179 @@
+    runAfter:
+    - unit-test
+    taskRef:
+      name: buildah
+
+```
+
+In this case, when you create pipelineRun you need to specify serviceAccount for every task.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunSpecs:
+    - pipelineTaskName: clone
+      taskServiceAccountName: git
+    - pipelineTaskName: unit-test
+      taskServiceAccountName: build
+    - pipelineTaskName: image-build
+      taskServiceAccountName: build
+```
+
+With the help of taskRunTemplate you will save a lot of work.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunSpecs:
+    - pipelineTaskName: clone
+      taskServiceAccountName: git
+  taskRunSpecTemplate:
+    taskServiceAccountName: build
+
+```
+
+Not only serviceAccount, but also data like computeResources metadata that you can specify for all tasks instead of specify them individually.
+
+## Design Details
+### taskRunTemplate
+Add taskRunSpecTemplate to PipelineRun.Spec so that user can specify configuration.
+```go
+// Add struct PipelineTaskRunTemplate
+type PipelineTaskRunTemplate struct {
+  TaskServiceAccountName string       `json:"taskServiceAccountName,omitempty"`
+  TaskPodTemplate        *PodTemplate `json:"taskPodTemplate,omitempty"`
+  // +listType=atomic
+  StepOverrides []TaskRunStepOverride `json:"stepOverrides,omitempty"`
+  // +listType=atomic
+  SidecarOverrides []TaskRunSidecarOverride `json:"sidecarOverrides,omitempty"`
+     
+  // +optional
+  Metadata *PipelineTaskMetadata `json:"metadata,omitempty"`
+     
+  // Compute resources to use for this TaskRun
+  ComputeResources *corev1.ResourceRequirements `json:"computeResources,omitempty"`
+}
+
+// Reference PipelineTaskRunTemplate in PipelineRunSpec as filed TaskRunSpecTemplate
+type PipelineRunSpec struct {
+	// +optional
+	PipelineRef *PipelineRef `json:"pipelineRef,omitempty"`
+	// +optional
+	PipelineSpec *PipelineSpec `json:"pipelineSpec,omitempty"`
+    .....
+    // TaskRunTemplate represent template of taskrun
++	// +optional
++	TaskRunTemplate PipelineTaskRunTemplate `json:"taskRunTemplate,omitempty"`
+}
+
+```
+
+Then, if you want to set a common configuration for all taskRun in pipelineRun, you can do so.
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunSpecs:
+    - pipelineTaskName: clone
+      taskServiceAccountName: git
+  taskRunTemplate:
+    taskServiceAccountName: build
+    metadata:
+      annotations:
+        vault.hashicorp.com/agent-inject-secret-foo: "/path/to/foo"
+        vault.hashicorp.com/role: role-name
+    taskPodTemplate:
+      nodeSelector:
+        disktype: ssd
+    computeResources:
+      requests:
+        cpu: 2
+    stepOverrides:
```

Comment:
would you mind removing `stepOverrides` from this example, as it can't be used along with `computeResources`?

Reply:
Thanks for your review @lbernick! 

I don't know why stepOverrides can not be used along with computeResources. Could you please help tell me the reason?  Thanks!  If so does this mean sidecarOverrides can only not be used along with computeResources either?

Reply:
It's explained in [TEP-0104](https://github.com/tektoncd/community/blob/main/teps/0104-tasklevel-resource-requirements.md); basically it's two different ways of configuring compute resources that aren't really compatible with each other. We are probably going to deprecate stepOverrides (see https://github.com/tektoncd/community/pull/786)